### ISSUE TITLE
Understanding benchmark results 

### ISSUE THREAD

**Issue by mslga** [etcd-io/etcd#10204]
Hello, 

I've installed etcd cluster v3.2.24 and encountered warning in log files:

```
2018-10-22 10:44:19.827715 W | etcdserver: read-only range request "key:\"/registry/serviceaccounts\" range_end:\"/registry/serviceaccountt\" count_only:true " with result "range_response_count:0 size:8" took too long (109.962962ms) to execute
2018-10-22 10:46:10.120222 I | mvcc: store.index: compact 1307252
2018-10-22 10:46:10.130271 I | mvcc: finished scheduled compaction at 1307252 (took 8.629232ms)
2018-10-22 10:46:14.387027 W | etcdserver: read-only range request "key:\"/registry/serviceaccounts\" range_end:\"/registry/serviceaccountt\" count_only:true " with result "range_response_count:0 size:8" took too long (158.582991ms) to execute
```

So, I've decided to run performance test as it described in: https://coreos.com/etcd/docs/latest/op-guide/performance.html
Test results:
```
benchmark --endpoints "https://etcd03:2379","https://etcd02:2379","https://etcd01:2379" --conns=1 --clients=1 range YOUR_KEY --consistency=l --total=10000
bench with linearizable range

Summary:
  Total:	27.9426 secs.
  Slowest:	0.0542 secs.
  Fastest:	0.0015 secs.
  Average:	0.0028 secs.
  Stddev:	0.0010 secs.
  Requests/sec:	357.8766

Response time histogram:
  0.0015 [1]	|
  0.0068 [9948]	|∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.0121 [40]	|
  0.0173 [6]	|
  0.0226 [2]	|
  0.0278 [0]	|
  0.0331 [0]	|
  0.0384 [1]	|
  0.0436 [0]	|
  0.0489 [1]	|
  0.0542 [1]	|

Latency distribution:
  10% in 0.0022 secs.
  25% in 0.0025 secs.
  50% in 0.0027 secs.
  75% in 0.0030 secs.
  90% in 0.0033 secs.
  95% in 0.0036 secs.
  99% in 0.0052 secs.
  99.9% in 0.0125 secs.

--------------------------------------------
benchmark --endpoints "https://etcd03:2379","https://etcd02:2379","https://etcd01:2379 --conns=1 --clients=1 range YOUR_KEY --consistency=s --total=10000
bench with serializable range

Summary:
  Total:	17.4235 secs.
  Slowest:	0.0374 secs.
  Fastest:	0.0009 secs.
  Average:	0.0017 secs.
  Stddev:	0.0008 secs.
  Requests/sec:	573.9365

Response time histogram:
  0.0009 [1]	|
  0.0045 [9936]	|∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.0082 [52]	|
  0.0118 [5]	|
  0.0155 [0]	|
  0.0191 [3]	|
  0.0228 [1]	|
  0.0265 [0]	|
  0.0301 [0]	|
  0.0338 [1]	|
  0.0374 [1]	|

Latency distribution:
  10% in 0.0011 secs.
  25% in 0.0014 secs.
  50% in 0.0016 secs.
  75% in 0.0020 secs.
  90% in 0.0023 secs.
  95% in 0.0027 secs.
  99% in 0.0038 secs.
  99.9% in 0.0085 secs.


----------------------------------------------
benchmark --endpoints "https://etcd03:2379","https://etcd02:2379","https://etcd01:2379 --conns=100 --clients=1000 range YOUR_KEY --consistency=l --total=100000

bench with linearizable range
Summary:
  Total:	19.7756 secs.
  Slowest:	1.0364 secs.
  Fastest:	0.0114 secs.
  Average:	0.1945 secs.
  Stddev:	0.1174 secs.
  Requests/sec:	5056.7259

Response time histogram:
  0.0114 [1]	|
  0.1139 [29332]	|∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.2164 [35917]	|∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.3189 [20506]	|∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.4214 [9263]	|∎∎∎∎∎∎∎∎∎∎
  0.5239 [3249]	|∎∎∎
  0.6264 [1217]	|∎
  0.7289 [350]	|
  0.8314 [137]	|
  0.9339 [26]	|
  1.0364 [2]	|

Latency distribution:
  10% in 0.0747 secs.
  25% in 0.1049 secs.
  50% in 0.1654 secs.
  75% in 0.2567 secs.
  90% in 0.3531 secs.
  95% in 0.4211 secs.
  99% in 0.5828 secs.
  99.9% in 0.7643 secs.
  
  
---------------------------------------------
benchmark --endpoints "https://etcd03:2379","https://etcd02:2379","https://etcd01:2379 --conns=100 --clients=1000 range YOUR_KEY --consistency=s --total=100000

bench with serializable range

Summary:
  Total:	6.7086 secs.
  Slowest:	0.4915 secs.
  Fastest:	0.0015 secs.
  Average:	0.0630 secs.
  Stddev:	0.0358 secs.
  Requests/sec:	14906.1722

Response time histogram:
  0.0015 [1]	|
  0.0505 [42075]	|∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.0995 [45737]	|∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.1485 [9389]	|∎∎∎∎∎∎∎∎
  0.1975 [2104]	|∎
  0.2465 [433]	|
  0.2955 [110]	|
  0.3445 [74]	|
  0.3935 [70]	|
  0.4425 [3]	|
  0.4915 [4]	|

Latency distribution:
  10% in 0.0271 secs.
  25% in 0.0393 secs.
  50% in 0.0558 secs.
  75% in 0.0781 secs.
  90% in 0.1057 secs.
  95% in 0.1283 secs.
  99% in 0.1844 secs.
  99.9% in 0.3175 secs.

---------------------------------------------
```
---------------------------------------------------
ETCD: v3.2.24
OS : Ubuntu 16.04
Hardware: RAM 8Gb, CPU 2 cores, 40 Gb HDD
---------------------------------------------------

Could You help me, please, to understand test results?

**Comment by mslga** [comment]
ETCD configuration:
```
# Environment file for etcd v3.2.24
ETCD_DATA_DIR=/var/lib/etcd
ETCD_ADVERTISE_CLIENT_URLS=https://192.168.xxx.xxx:2379
ETCD_INITIAL_ADVERTISE_PEER_URLS=https://192.168.xxx.xxx:2380
ETCD_INITIAL_CLUSTER_STATE=existing
ETCD_METRICS=basic
ETCD_LISTEN_CLIENT_URLS=https://192.168.xxx.xxx:2379,https://127.0.0.1:2379
ETCD_ELECTION_TIMEOUT=5000
ETCD_HEARTBEAT_INTERVAL=250
ETCD_INITIAL_CLUSTER_TOKEN=k8s_etcd
ETCD_LISTEN_PEER_URLS=https://192.168.xxx.xxx:2380
ETCD_NAME=etcd3
ETCD_PROXY=off
ETCD_INITIAL_CLUSTER=etcd1=https://192.168.xxx.xxx:2380,etcd2=https://192.168.xxx.yyy:2380,etcd3=https://192.168.xxx.zzz:2380
ETCD_AUTO_COMPACTION_RETENTION=8
ETCD_SNAPSHOT_COUNT=10000

# TLS settings
ETCD_TRUSTED_CA_FILE=/etc/ssl/etcd/ssl/ca.pem
ETCD_CERT_FILE=/etc/ssl/etcd/ssl/etcd03.pem
ETCD_KEY_FILE=/etc/ssl/etcd/ssl/etcd03-key.pem
ETCD_CLIENT_CERT_AUTH=true

ETCD_PEER_TRUSTED_CA_FILE=/etc/ssl/etcd/ssl/ca.pem
ETCD_PEER_CERT_FILE=/etc/ssl/etcd/ssl/etcd03.pem
ETCD_PEER_KEY_FILE=/etc/ssl/etcd/ssl/etcd03-key.pem
ETCD_PEER_CLIENT_CERT_AUTH=True
```

**Comment by wenjiaswe** [comment]
First, just a friendly reminder that https://coreos.com/etcd/docs/latest/op-guide/performance.html this link mentioned in your post is deprecated, as marked on top of the webpage:

>  "These docs are deprecated while they are being migrated to Red Hat. For the most up to date docs, please see the corresponding GitHub repository."

Please refer to https://github.com/etcd-io/etcd/blob/688043a7c2ac8bbb0e73ca5694c7815275865e24/Documentation/op-guide/performance.md.

Now back to your question, would you please be more specific about what aspect are you trying to understand?

**Comment by mslga** [comment]
Thanks for the link.
 
I understood the messages from the log so that my servers have problems with I/O performance. Therefore I've launched tests.
But I don't know which indicators I should pay attention to. And which of the indicators in my test are "bad" and indicate a performance degradation.
 
I've not found information which indicators can be "good" and which I could use as a comparison.
For example, is the indicator "Latency distribution: 99.9% in 0.7643 secs." bad? 

And if it is, what should I do? What can you advise in this case?

**Comment by hexfusion** [comment]
@mslga per [FAQ :system-requirements ](https://github.com/etcd-io/etcd/blob/master/Documentation/faq.md#system-requirements) SSD is strongly recommended.

> I've not found information which indicators can be "good" and which I could use as a comparison.
For example, is the indicator "Latency distribution: 99.9% in 0.7643 secs." bad?

If you review the performace doc provided vy @wenjiaswe you will see with SSD the same lineralized workload you ran avg latency is 5.5ms yours on HDD is 764.3ms. 

> 2018-10-22 10:46:14.387027 W | etcdserver: read-only range request "key:\"/registry/serviceaccounts\" range_end:\"/registry/serviceaccountt\" count_only:true " with result "range_response_count:0 size:8" took too long (158.582991ms) to execute

More I/O related issues please review [faq: what-does-the-etcd-warning-apply-entries-took-too-long-mean](https://github.com/etcd-io/etcd/blob/master/Documentation/faq.md#what-does-the-etcd-warning-apply-entries-took-too-long-mean)

> And if it is, what should I do? What can you advise in this case?

Basically you need ot update to SSD or you will contantly be running into this I/O issue. If that is not possible making sure your HDD which stores data-dir is decicated to this only and look at fine tuning your [compaction](https://github.com/etcd-io/etcd/blob/ae25c5e1320f731a2ffaafbf756aca8b0a94dfab/Documentation/op-guide/maintenance.md#history-compaction-v3-api-key-value-database) settings.
